{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: appnope==0.1.4 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 1)) (0.1.4)\n",
      "Requirement already satisfied: APScheduler==3.10.4 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 2)) (3.10.4)\n",
      "Requirement already satisfied: asttokens==2.4.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 3)) (2.4.1)\n",
      "Requirement already satisfied: attrs==23.2.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 4)) (23.2.0)\n",
      "Requirement already satisfied: certifi==2024.2.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 5)) (2024.2.2)\n",
      "Requirement already satisfied: charset-normalizer==3.3.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 6)) (3.3.2)\n",
      "Requirement already satisfied: comm==0.2.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 7)) (0.2.2)\n",
      "Requirement already satisfied: debugpy==1.8.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 8)) (1.8.1)\n",
      "Requirement already satisfied: decorator==5.1.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 9)) (5.1.1)\n",
      "Requirement already satisfied: exceptiongroup==1.2.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 10)) (1.2.0)\n",
      "Requirement already satisfied: executing==2.0.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 11)) (2.0.1)\n",
      "Requirement already satisfied: h11==0.14.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 12)) (0.14.0)\n",
      "Requirement already satisfied: idna==3.6 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 13)) (3.6)\n",
      "Requirement already satisfied: importlib_metadata==7.0.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 14)) (7.0.2)\n",
      "Requirement already satisfied: ipykernel==6.29.3 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 15)) (6.29.3)\n",
      "Requirement already satisfied: ipython==8.18.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 16)) (8.18.1)\n",
      "Requirement already satisfied: ipywidgets==8.1.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 17)) (8.1.2)\n",
      "Requirement already satisfied: jedi==0.19.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 18)) (0.19.1)\n",
      "Requirement already satisfied: jupyter_client==8.6.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 19)) (8.6.1)\n",
      "Requirement already satisfied: jupyter_core==5.7.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 20)) (5.7.2)\n",
      "Requirement already satisfied: jupyterlab_widgets==3.0.10 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 21)) (3.0.10)\n",
      "Requirement already satisfied: markdown-it-py==3.0.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 22)) (3.0.0)\n",
      "Requirement already satisfied: matplotlib-inline==0.1.6 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 23)) (0.1.6)\n",
      "Requirement already satisfied: mdurl==0.1.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 24)) (0.1.2)\n",
      "Requirement already satisfied: nest-asyncio==1.6.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 25)) (1.6.0)\n",
      "Requirement already satisfied: outcome==1.3.0.post0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 26)) (1.3.0.post0)\n",
      "Requirement already satisfied: packaging==24.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 27)) (24.0)\n",
      "Requirement already satisfied: parso==0.8.3 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 28)) (0.8.3)\n",
      "Requirement already satisfied: pexpect==4.9.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 29)) (4.9.0)\n",
      "Requirement already satisfied: platformdirs==4.2.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 30)) (4.2.0)\n",
      "Requirement already satisfied: prompt-toolkit==3.0.43 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 31)) (3.0.43)\n",
      "Requirement already satisfied: psutil==5.9.8 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 32)) (5.9.8)\n",
      "Requirement already satisfied: ptyprocess==0.7.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 33)) (0.7.0)\n",
      "Requirement already satisfied: pure-eval==0.2.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 34)) (0.2.2)\n",
      "Requirement already satisfied: Pygments==2.17.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 35)) (2.17.2)\n",
      "Requirement already satisfied: PySocks==1.7.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 36)) (1.7.1)\n",
      "Requirement already satisfied: python-dateutil==2.9.0.post0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 37)) (2.9.0.post0)\n",
      "Requirement already satisfied: python-dotenv==1.0.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 38)) (1.0.1)\n",
      "Requirement already satisfied: pytz==2024.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 39)) (2024.1)\n",
      "Requirement already satisfied: pyzmq==25.1.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 40)) (25.1.2)\n",
      "Requirement already satisfied: requests==2.31.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 41)) (2.31.0)\n",
      "Requirement already satisfied: rich==13.7.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 42)) (13.7.1)\n",
      "Requirement already satisfied: selenium==4.18.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 43)) (4.18.1)\n",
      "Requirement already satisfied: six==1.16.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 44)) (1.16.0)\n",
      "Requirement already satisfied: sniffio==1.3.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 45)) (1.3.1)\n",
      "Requirement already satisfied: sortedcontainers==2.4.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 46)) (2.4.0)\n",
      "Requirement already satisfied: stack-data==0.6.3 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 47)) (0.6.3)\n",
      "Requirement already satisfied: tornado==6.4 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 48)) (6.4)\n",
      "Requirement already satisfied: traitlets==5.14.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 49)) (5.14.2)\n",
      "Requirement already satisfied: trio==0.24.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 50)) (0.24.0)\n",
      "Requirement already satisfied: trio-websocket==0.11.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 51)) (0.11.1)\n",
      "Requirement already satisfied: typing_extensions==4.10.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 52)) (4.10.0)\n",
      "Requirement already satisfied: tzlocal==5.2 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 53)) (5.2)\n",
      "Requirement already satisfied: urllib3==2.2.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 54)) (2.2.1)\n",
      "Requirement already satisfied: wcwidth==0.2.13 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 55)) (0.2.13)\n",
      "Requirement already satisfied: webdriver-manager==4.0.1 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 56)) (4.0.1)\n",
      "Requirement already satisfied: widgetsnbextension==4.0.10 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 57)) (4.0.10)\n",
      "Requirement already satisfied: wsproto==1.2.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 58)) (1.2.0)\n",
      "Requirement already satisfied: zipp==3.18.0 in ./.venv/lib/python3.9/site-packages (from -r requirements.txt (line 59)) (3.18.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.service import Service\n",
    "from webdriver_manager.chrome import ChromeDriverManager\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.common.exceptions import NoSuchElementException\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "\n",
    "from apscheduler.schedulers.background import BackgroundScheduler\n",
    "\n",
    "from contextlib import contextmanager\n",
    "\n",
    "from rich.console import Console\n",
    "from rich.table import Table\n",
    "from rich.live import Live\n",
    "\n",
    "from getpass import getpass\n",
    "\n",
    "from datetime import datetime, timezone\n",
    "import re\n",
    "import random\n",
    "import time\n",
    "\n",
    "console = Console()\n",
    "\n",
    "ADVENTURES = 'adventures'\n",
    "RAIDS = 'raids'\n",
    "logs = {\n",
    "    ADVENTURES: '',\n",
    "    RAIDS: '',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' scheduler init. '''\n",
    "\n",
    "def calc_new_interval_between(x, y):\n",
    "    return random.uniform(x, y)\n",
    "\n",
    "def cleanup_scheduler(scheduler):\n",
    "    if scheduler.running:\n",
    "        print(\"Shutting down scheduler...\")\n",
    "        scheduler.shutdown(wait=True)\n",
    "    else:\n",
    "        print(\"Scheduler already stopped.\")\n",
    "\n",
    "# Context manager for BackgroundScheduler\n",
    "@contextmanager\n",
    "def managed_scheduler(*args, **kwargs):\n",
    "    scheduler = BackgroundScheduler(*args, **kwargs)\n",
    "    scheduler.start()\n",
    "    try:\n",
    "        yield scheduler\n",
    "    finally:\n",
    "        cleanup_scheduler(scheduler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#             roman,      roman,   hun,       gaul,     egyptian,     teuton\n",
    "usernames = ['robot723', 'athos', 'porthos', 'aramis', 'd.artagnan', 'richelieu']\n",
    "passwords = ['Robot6210!', 'Athos2024!', 'Porthos2024!', 'Aramis2024!', 'dArtagnan2024!', 'Richelieu2024!']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' Open Travian International 2 server and login if not already '''\n",
    "\n",
    "def attempt_login(site, username, password, driver):\n",
    "    driver.get(site)\n",
    "\n",
    "    try:\n",
    "        input_username = driver.find_element(By.XPATH, '//*[@id=\"loginForm\"]/tbody/tr[1]/td[2]/input')\n",
    "        input_pwd = driver.find_element(By.XPATH, '//*[@id=\"loginForm\"]/tbody/tr[2]/td[2]/input')\n",
    "        button_login = driver.find_element(By.CSS_SELECTOR, 'button[type=\"submit\"][value=\"Login\"].textButtonV1.green')\n",
    "\n",
    "        input_username.send_keys(username)\n",
    "        input_pwd.send_keys(password)\n",
    "        button_login.click()\n",
    "    except NoSuchElementException:\n",
    "        print(\"User already signed in!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' Send hero on adventure '''\n",
    "\n",
    "def attempt_to_start_adventure(driver, scheduler):\n",
    "    button_hero_status = driver.find_element(By.XPATH, '//*[@id=\"topBarHero\"]/div/a/i')\n",
    "\n",
    "    # dynamic ID, using href to reference\n",
    "    button_adventures = driver.find_element(By.XPATH, '//a[contains(@href, \"/hero/adventures\")]')\n",
    "    try:\n",
    "        num_adventures = button_adventures.find_element(By.XPATH, \"./div\")\n",
    "    except NoSuchElementException:\n",
    "        num_adventures = None\n",
    "\n",
    "    if num_adventures and 'heroHome' in button_hero_status.get_attribute('class') and int(num_adventures.text) > 0:\n",
    "        button_adventures.click()\n",
    "\n",
    "        button_start_first_adventure = driver.find_element(By.XPATH, '//*[@id=\"heroAdventure\"]/table/tbody/tr[1]/td[5]/button')\n",
    "        button_start_first_adventure.click()\n",
    "\n",
    "        button_continue = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.XPATH, '//*[@id=\"heroAdventure\"]/div/button')))\n",
    "        button_continue.click()\n",
    "\n",
    "        logs[ADVENTURES] = \"Successfully sent out hero on adventure!\"\n",
    "    elif num_adventures is None:\n",
    "        logs[ADVENTURES] = \"No adventures found.\"\n",
    "    elif 'heroRunning' in button_hero_status.get_attribute('class'):\n",
    "        logs[ADVENTURES] = \"Hero is already on an adventure!\"\n",
    "    else:\n",
    "        logs[ADVENTURES] = \"Error attempting to start adventure.\"\n",
    "    \n",
    "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), id=ADVENTURES, args=[driver, scheduler], replace_existing=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' Upgrade lowest level field '''\n",
    "\n",
    "def attempt_to_upgrade_field(driver):\n",
    "    button_resource_fields_page = driver.find_element(By.XPATH, '//*[@id=\"navigation\"]/a[1]')\n",
    "    button_resource_fields_page.click()\n",
    "\n",
    "    resource_field_container = driver.find_element(By.XPATH, '//*[@id=\"resourceFieldContainer\"]')\n",
    "    resource_field_links = resource_field_container.find_elements(By.TAG_NAME, 'a')\n",
    "\n",
    "    # add all fields that we can upgrade to new list\n",
    "    upgradable_fields = []\n",
    "    for link in resource_field_links:\n",
    "        if 'good' in link.get_attribute('class'):\n",
    "            upgradable_fields.append(link)\n",
    "\n",
    "    # find the lowest level field\n",
    "    lowest_level_field = None\n",
    "    lowest_level = 999\n",
    "    for field in upgradable_fields:\n",
    "        # do not upgrade fields that are already being upgraded\n",
    "        if 'underConstruction' in field.get_attribute('class'): continue\n",
    "\n",
    "        div_curr_level = field.find_element(By.XPATH, './div')\n",
    "        curr_level = int(div_curr_level.text.strip() or 0)\n",
    "        if curr_level < lowest_level:\n",
    "            lowest_level_field = field\n",
    "            lowest_level = curr_level\n",
    "\n",
    "    if lowest_level_field:\n",
    "        lowest_level_field.click()\n",
    "\n",
    "        button_upgrade = driver.find_element(By.CSS_SELECTOR, '.textButtonV1.green.build')\n",
    "        button_upgrade.click()\n",
    "    else:\n",
    "        print(\"No upgradable fields were found!\")\n",
    "\n",
    "    #scheduler.add_job(attempt_to_upgrade_field, 'interval', seconds=calc_new_interval_between(58, 307), replace_existing=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def activate_farm_list_raids_for(id, driver):\n",
    "    name = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.XPATH, f'{id}/div/div[1]/div[2]/div[1]'))).text\n",
    "\n",
    "    table = driver.find_element(By.XPATH, f'{id}/div/div[2]/table')\n",
    "    table_body = table.find_element(By.XPATH, './tbody')\n",
    "    table_footer = table.find_element(By.XPATH, './tfoot')\n",
    "    table_rows = table_body.find_elements(By.TAG_NAME, 'tr')\n",
    "\n",
    "    for row in table_rows:\n",
    "        try:\n",
    "            checkbox = row.find_element(By.XPATH, './td[1]/label/input')\n",
    "        except NoSuchElementException:\n",
    "            continue\n",
    "\n",
    "        last_raid_state_container = row.find_element(By.XPATH, './td[7]')\n",
    "        try:\n",
    "            last_raid_state = last_raid_state_container.find_element(By.TAG_NAME, 'i')\n",
    "        except NoSuchElementException:\n",
    "            last_raid_state = None\n",
    "        \n",
    "        curr_state_container = row.find_element(By.XPATH, './td[2]')\n",
    "        try:\n",
    "            curr_state = curr_state_container.find_element(By.TAG_NAME, 'i')\n",
    "        except NoSuchElementException:\n",
    "            curr_state = None\n",
    "\n",
    "        num_troops_container = row.find_element(By.XPATH, './td[6]')\n",
    "        num_troops = int(num_troops_container.find_element(By.XPATH, './div[1]/span[1]/span[1]').text)\n",
    "\n",
    "        try:\n",
    "            filthy_nominator   = table_footer.find_element(By.XPATH, './tr[1]/td[2]/div/div/span/span/span[1]').text\n",
    "            filthy_denominator = table_footer.find_element(By.XPATH, './tr[1]/td[2]/div/div/span/span/span[2]').text\n",
    "            nominator   = int(re.sub(r'[^\\d]', '', filthy_nominator))\n",
    "            denominator = int(re.sub(r'[^\\d]', '', filthy_denominator))\n",
    "        except NoSuchElementException:\n",
    "            nominator = denominator = None\n",
    "\n",
    "        if (not curr_state or 'attack_small' not in curr_state.get_attribute('class')) and \\\n",
    "        (not last_raid_state or 'attack_won_withoutLosses_small' in last_raid_state.get_attribute('class')) and \\\n",
    "        ((not nominator and not denominator) or nominator + num_troops <= denominator):\n",
    "            checkbox.click()\n",
    "\n",
    "    # start raids\n",
    "    button_start_raids = driver.find_element(By.XPATH, f'{id}/div/div[1]/button')\n",
    "    button_start_raids.click()\n",
    "\n",
    "    logs[RAIDS] = f'Finished raid logic for {name}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def send_troops_to_farm(driver, scheduler):\n",
    "    if driver.current_url != 'https://ts2.x1.international.travian.com/build.php?id=39&gid=16&tt=99':\n",
    "        button_buildings_page = driver.find_element(By.XPATH, '//*[@id=\"navigation\"]/a[2]')\n",
    "        button_buildings_page.click()\n",
    "\n",
    "        button_rally_point = driver.find_element(By.XPATH, '//*[@id=\"villageContent\"]/div[21]/a')\n",
    "        button_rally_point.click()\n",
    "\n",
    "        button_farm_list = driver.find_element(By.XPATH, '//a[contains(@href, \"/build.php?id=39&gid=16&tt=99\")]')\n",
    "        button_farm_list.click()\n",
    "\n",
    "    oases    = '//*[@id=\"rallyPointFarmList\"]/div[2]/div[2]'\n",
    "    activate_farm_list_raids_for(oases, driver)\n",
    "\n",
    "    villages = '//*[@id=\"rallyPointFarmList\"]/div[2]/div[3]'\n",
    "    activate_farm_list_raids_for(villages, driver)\n",
    "\n",
    "    scheduler.add_job(send_troops_to_farm, 'interval', seconds=calc_new_interval_between(183, 302), id=RAIDS, args=[driver, scheduler], replace_existing=True)\n",
    "\n",
    "    logs[RAIDS] = \"Finished raid attempt.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_job_scheduler_table_from(scheduler):\n",
    "    table = Table(title=\"Scheduled Jobs\")\n",
    "\n",
    "    table.add_column(\"Job ID\", style=\"cyan\", no_wrap=True)\n",
    "    table.add_column(\"Next Run At\", style=\"magenta\")\n",
    "    table.add_column(\"Countdown\", style=\"green\")\n",
    "    table.add_column(\"Log\", style=\"blue\")\n",
    "\n",
    "    now = datetime.now(timezone.utc)\n",
    "    for job in scheduler.get_jobs():\n",
    "        next_run = job.next_run_time\n",
    "        countdown = (next_run - now).total_seconds()\n",
    "        table.add_row(job.id, str(next_run), str(int(countdown)), logs[job.id])\n",
    "    \n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b466394313a542659af235710cf92dcf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:05 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:05 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:06 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:06 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:07 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:07 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:08 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:08 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:09 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:09 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:10 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:10 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:11 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Job \"attempt_to_start_adventure (trigger: interval[0:00:01], next run at: 2024-03-13 18:07:11 MDT)\" raised an \n",
       "exception\n",
       "Traceback (most recent call last):\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/executors/base.py\", \n",
       "line 125, in run_job\n",
       "    retval = job.func(*job.args, **job.kwargs)\n",
       "  File \"/var/folders/tf/s_38qxxn53jg20v0_kpkc6580000gp/T/ipykernel_48106/1547744773.py\", line 30, in \n",
       "attempt_to_start_adventure\n",
       "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=calc_new_interval_between(307, 902), \n",
       "id=ADVENTURES, replace_existing=True)\n",
       "  File \n",
       "\"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/schedulers/base.py\", \n",
       "line 448, in add_job\n",
       "    job = Job(self, **job_kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "49, in __init__\n",
       "    self._modify(id=id or uuid4().hex, **kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/job.py\", line \n",
       "180, in _modify\n",
       "    check_callable_args(func, args, kwargs)\n",
       "  File \"/Users/danielgarcia/SideProjects/TravianAuto/.venv/lib/python3.9/site-packages/apscheduler/util.py\", line \n",
       "372, in check_callable_args\n",
       "    raise ValueError('The following arguments have not been supplied: %s' %\n",
       "ValueError: The following arguments have not been supplied: driver, scheduler\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shutting down scheduler...\n"
     ]
    }
   ],
   "source": [
    "''' web driver init. '''\n",
    "def init_webdriver():\n",
    "    options = webdriver.ChromeOptions()\n",
    "    options.add_argument(\"start-maximized\")\n",
    "    return webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=options)\n",
    "\n",
    "''' init. and login to server '''\n",
    "input_site = 'https://ts2.x1.international.travian.com/dorf1.php'\n",
    "#input_site = input(\"Site:\") # Commented out while testing\n",
    "input_username = input(\"Username:\")\n",
    "input_password = getpass(\"Password:\")\n",
    "\n",
    "driver = init_webdriver()\n",
    "attempt_login(input_site, input_username, input_password, driver)\n",
    "\n",
    "''' begin scheduler tasks '''\n",
    "with managed_scheduler() as scheduler:\n",
    "    scheduler.add_job(attempt_to_start_adventure, 'interval', seconds=1, id=ADVENTURES, args=[driver, scheduler])\n",
    "    scheduler.add_job(send_troops_to_farm, 'interval', seconds=20, id=RAIDS, args=[driver, scheduler])\n",
    "\n",
    "    with Live(generate_job_scheduler_table_from(scheduler), refresh_per_second=1, console=console) as live:\n",
    "        try:\n",
    "            while True:\n",
    "                live.update(generate_job_scheduler_table_from(scheduler))\n",
    "                time.sleep(1)\n",
    "        except (KeyboardInterrupt, SystemExit):\n",
    "            pass"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
